rfd/0173-terraform-machine-id.md

```diff
@@ -0,0 +1,303 @@
+  - if the call fails because of missing permissions, output a user-friendly error such as:
+    ```
+    Failed to create bootstrap resources using your local credentials (user "hugo.hervieux@goteleport.com", address "mytenant.telpeort.sh:443").
+    Please chech you have the rights to create role, bot and token resources. If you got recently granted those rights you might need to re-log in.
+    (tsh logout --proxy="mytenant.teleport.sh:443" --user="hugo.hervieux@goteleport.com")
+    ```
+- continue the provider flow with the following onboarding config
+  ```hcl
+  onboarding = {
+    token = "<secret token that got generated earlier>"
+    join_method = "token"
+  }
+  ```
+
+#### Backward compatibility
+
+By default, when neither `bootstrap` nor `onboarding` values are set the provided uses the existing `identity_*` values.
+This ensures compatibility with existing setups.
+
+`onboarding`, `bootstrap` and the `identity_` settings are mutually exclusive and the Provider will refuse to start if
+more than one is set. This will avoid any un-tested hybrid configuration.
+
+The error message would look like:
+```
+Invalid provider configuration. `bootstrap`, `onboarding` and `identity_*`/`key_*`/`profile_*` values are mutually exclusive.
+You must set only one.
+
+- `bootstrap` is used when running Terraform on a developper laptop
+- `onboarding` is used when running Terraform in CI/CD pipelines such as GitHub Actions or GitLab CI.
+- passing certificates directly with `identity_*`, `key_*` or `profile_*` is used when you already have Teleport credentials for the provider.
+```
+
+### Security
+
+#### Benefits
+
+This approach makes using MachineID and adopting short-lived certificates easier, especially in CI/CD pipelines.
+Switching to short-lived certificates and delegated join methods improves security as there's less material to
+exfiltrate and users can fine-tune the token permissions (allow joining based on the service account/github
+project/workload location). Adopting "MFA for admin" will also be easier for existing users.
+
+This approach also improves security by not writing to disk the MachineID-generated certs. In case of
+misconfigured/permissive ACLs, an attacker already present on the host will not be able to obtain certs from reading the
+filesystem. Exfiltrating the MachineID certs requires dumping the process memory.
+
+#### Risks
+
+The main risk is the amount of resources created by the `bootstrap` configuration. Each terraform invocation will
+create a bot resource. Too many resource could create noise or affect Teleport's performance. This risk is mitigated
+by:
+- the fact `bootstrap` is only usable on local laptop. It requires a valid `~/.tsh` profile and the ability to pass an
+  MFA challenge (when MFA4A is enabled, which we are pushing everywhere and is Cloud's default). Intensive CI usage will
+  use existing bot resources and the `onboarding` configuration.
+- the bootstrap resource expiry, by default 1 hour.
```

Comment:
Am I reading this correctly as the bots from `bootstrap` being removed after 1 hour, not just the local certs having a 1hr TTL?

I know when I'm first trying a new thing out with Terraform, or even first trying out a new IaC tool (and for some small number of people we may be the first thing they use Terraform with), I can easily end up running `terraform apply` several times per minute to _get this freaking thing working_, but I think an hour is a good default. I probably wouldn't create _more_ than ~80 bots during that time. If this is just the local TTL though, and the bots hang around on the backend, this is my notification that I could easily create ~80 bots in an hour. Several days in a row.

Reply:
You're correct, this si the resoucre TTL, not the certs. So I don't expect more than a 100 bots in a very intensive setup session. This is noisy, but should not impact performance. We can add a GET call beforeto count the bots and limit their number if needed (e.g. 20 bots max)